internal/emscripten/emscripten.go

```diff
@@ -89,8 +110,60 @@ func (v *InvokeFunc) Call(ctx context.Context, mod api.Module, stack []uint64) {
 		panic(err)
 	}
 
-	err = f.CallWithStack(ctx, stack)
+	// The Go implementation below mimics the Emscripten JS behaviour to support
+	// longjmps from indirect function calls. The implementation of these
+	// indirection function calls in Emscripten JS is like this:
+	//
+	// function invoke_iii(index,a1,a2) {
+	//  var sp = stackSave();
+	//  try {
+	//    return getWasmTableEntry(index)(a1,a2);
+	//  } catch(e) {
+	//    stackRestore(sp);
+	//    if (e !== e+0) throw e;
+	//    _setThrew(1, 0);
+	//  }
+	//}
+
+	// This is the equivalent of "var sp = stackSave();".
+	// We reuse savedStack to save allocations.
+	var savedStack [2]uint64
+	err = mod.ExportedFunction("stackSave").CallWithStack(ctx, savedStack[:])
 	if err != nil {
 		panic(err)
 	}
+
+	err = f.CallWithStack(ctx, stack)
+	if err != nil {
+		// Module closed: any calls will just fail with the same error.
+		if _, ok := err.(*sys.ExitError); ok {
+			panic(err)
+		}
+
+		// This is the equivalent of "stackRestore(sp);".
+		// Do not overwrite err here to preserve the original error.
+		if err := mod.ExportedFunction("stackRestore").CallWithStack(ctx, savedStack[:]); err != nil {
+			panic(err)
+		}
+
+		// If we encounter ThrowLongjmpError, this means that the C code did a
+		// longjmp, which in turn called _emscripten_throw_longjmp and that is
+		// a host function that panics with ThrowLongjmpError. In that case we
+		// ignore the error because we have restored the stack to what it was
+		// before the indirect function call, so the program can continue.
+		// This is the equivalent of the "if (e !== e+0) throw e;" line in the
+		// JS implementation, which checks if the error is Infinity, which is
```

Comment:
which checks if the error is not a number, which is what the JS implementation throws (Infinity for _emscripten_throw_longjmp, a memory address for C++ exceptions).

Reply:
Changed :) Also looked into implementing C++ exceptions but that looks a lot more complex.